ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -151,19 +156,15 @@ void onNeighboursPacketReceived(
 
   private void queryNearestNodes() {
     final int concurrentRequestLimit = 3;
-    if (outstandingRequestList.isEmpty()) {
+    if (outstandingNeighboursRequestList.isEmpty()) {
       final List<DiscoveryPeer> queryCandidates =
           determineFindNodeCandidates(concurrentRequestLimit);
```

Comment:
Adding some notes based on our offline discussion:

It looks to me like the algorithm for selecting nodes to query (defined in `determineFindNodeCandidates`) is not quite right.  The [docs](https://github.com/ethereum/devp2p/blob/master/discv4.md#recursive-lookup) describe how to select the next set of nodes to query:
> Of the k nodes the initiator has heard of closest to the target, it picks α that it has not yet queried and resends FindNode to them.

As it stands currently, it looks like you're always selecting the peers closest to the target regardless of whether we have contacted those peers already, or if those peers are contained in the set of the `k` closest peers.